### PR TITLE
Allow static input mappings

### LINKED ISSUE #8733: Simplify injection of static values via `zeebe:input`
**Is your feature request related to a problem? Please describe.**

As a user I'm happily automating my processes with Zeebe. What I already understood is that Zeebe has a clear way of distinguishing expressions (`= someVariable`) from static values (strings, `HELLO WORLD`).

What I also understood is that headers shall be used for "static configuration". Inputs on the other hand shall be used to (dynamic) variable binding.

As a matter of fact though variables can also be of "static" nature. I.e. as a user I build an email worker, require the `PORT` to be passed in as a variable. I do not want to have that variable in my process; I simply want to set it up in the easiest possible manner and want my team to use it.

So in order to provide that static `PORT=25` value I need to specify the input using a `=25` expression. What I wanted to do instead is to simply enter `25` (what would pass the result to my worker as the string `"25"`. Intuitively I thought that simple conversion was supported, [it is for messages](https://docs.camunda.io/docs/components/modeler/bpmn/message-events/message-events/#messages).

image

_Camunda Modeler + example diagram._

Bottom line: As zeebe already has a clear understanding if I pass an expression `= ABC` or not (not prefixed with `=`) it does not make sense why it would not be forgiving to me as a user.

**Describe the solution you'd like**

The engine uses `=` to determine if an input is an expression, if `INPUT` is not prefixed with `=` it will simply provide `INPUT` as a string.

In other places, i.e. [messages](https://docs.camunda.io/docs/components/modeler/bpmn/message-events/message-events/#messages) it exactly works like this.

**Describe alternatives you've considered**

In a world where I as a user have full understanding of zeebe internals and/or can clearly distinguish I can freely switch between inputs and headers; so accepting the situation would be an option.

In the world of element templates and re-usable worker implementations (i.e. email or rest task) we'd always need to prefix everything, even plain strings with `=`. It does make things explicit but also requires users to learn more things as they get started automating with zeebe.

**Additional context**

This came up in the context of [improved feel expression editing support](https://github.com/camunda/product-hub/issues/33) as well as [connectors](https://github.com/camunda/product-hub/issues/18).

### PR DESCRIPTION
**Pull request by menski**
## Description

Before input mappings always had to be expressions. It is not clear anymore why we enforced this, but Zeebe already supports static expressions based on the value of the field.

This change remove the validation for non static expression on input mappings to improve the user expierence we can offer in the [connectors use case](https://github.com/camunda/team-connectors/issues/85).

The static value will be escaped as a java string, and not as a JSON string as it would otherwise additionally escape `/` characters which will be used for URL parameters and additionally escaping them decreases the user experience.

The current implementation might multiple times escape characters like `"` and `\`, as they are escaped initially to be given to the FEEL engine and later when retrieving the value from the variable record.

## Related issues

closes #8733 

Documentation issue: https://github.com/camunda-cloud/camunda-cloud-documentation/issues/643

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: ❌ I think this shows why static values are not supported so far.

All other places allow a static value to be defined as a string without the `"` double quotes around it. However, here you need the quotes. When you omit the quotes, the static value will be considered the name of a variable.

Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: Note that this differs from `jobType`, `assignee`, `candidateGroups`, `error-name` and a few others.

Comment (menski) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: I see your point, let me look into it

Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: 💡 You might have a look at 'escaping' the result of the source expression if it is an instanceof StaticExpression.

Comment (nikku) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: The idea of the initial issue (https://github.com/camunda-cloud/zeebe/issues/8733) was to auto-wrap all non-expression inputs (`= 12`) as strings (`12` becomes `= "12"`).

Interesting case is escaping `"` inside the string, i.e. handling `My name is "walt"`.

Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/ActivityTest.java:72]: I agree that we should handle the static values as strings :+1: 

Static values don't have type information. We need to treat them as strings.   

Regarding the escaping of `"` in a static value, is an interesting edge case. We should create a test case for it.

**Review by saig0 — DISMISSED** [review]
@menski nice try :sweat_smile: :+1: 

As already mentioned in the comments, we should handle the static values as strings.

**Comment by menski** [comment]
Thanks all for the reviews and comments. I added a commit which adds escaping the static expression output before putting it in the input mapping context expression. This works well for simple cases, i.e. `text`, `12` and `true`. But when I added tests for the edge case like `my name is "walt"` it seems that the value of the resulting variable is "over escaped", see the output of the failing test case below

```
java.lang.AssertionError: 
Expecting map:
  {"boolean"=""true"", "list"=""[{\\\"key\\\": \\\"value\\\"}, 2, true]"", "number"=""1"", "object"=""{\\\"key\\\": \\\"value\\\"}"", "quotedText"=""my name is \\\"walt\\\""", "text"=""text"", "url"=""https://my.api.com/{{orgId}}/{{userId}}""}
to contain:
  [MapEntry[key="text", value=""text""],
    MapEntry[key="url", value=""https://my.api.com/{{orgId}}/{{userId}}""],
    MapEntry[key="quotedText", value=""my name is \"walt\"""],
    MapEntry[key="object", value=""{\"key\": \"value\"}""],
    MapEntry[key="list", value=""[{\"key\": \"value\"}, 2, true]""],
    MapEntry[key="number", value=""1""],
    MapEntry[key="boolean", value=""true""]]
but could not find the following map entries:
  [MapEntry[key="quotedText", value=""my name is \"walt\"""],
    MapEntry[key="object", value=""{\"key\": \"value\"}""],
    MapEntry[key="list", value=""[{\"key\": \"value\"}, 2, true]""]]
```

@korthout any idea where in the engine this additional escaping of the `/` slash could happen, I tried to debug it but could find the moment it is added. It could also be that I don't understand escaping :laughing:

**Comment by korthout** [comment]
> any idea where in the engine this additional escaping of the `\` slash could happen

@menski Interesting situation. It happens in `VariableRecord.getValue` ([code](https://github.com/camunda-cloud/zeebe/blob/main/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/variable/VariableRecord.java#L43-L46)). Because it uses `MsgPackConverter.convertToJson` which escapes it as well. 

The thing is that you need to escape the expression because it will be evaluated by FEEL. But it's also escaped when we read the value from the DirectBuffer. For some time it was still unclear to me is where we actually call this getValue method, but it's used indirectly when you call the `ProcessInstances.getCurrentVariables` in your test. 

What is strange to me is that writing the value of a `VariableRecord` with `setValue` is just raw, while reading it with the `getValue` is using convertToJson. 

It leaves some questions to be answered. Is the test wrong for reading it like this? Perhaps we need a way to read the value of the variable 'raw'. Or perhaps this is all correct and we should simply accept a double escaping (once for feel and once for reading the variable record's value). WDYT?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:158]: 🔧 the feel syntax is closer to JSON than it is to Java. I recommend escaping it for JSON usage.

```suggestion
          expression = String.format("\"%s\"", StringEscapeUtils.escapeJson(expression));
```

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:155]: 🔧 local variable expression can be final by splitting the declaration and assignment by using an else statement in the if below to fit the assignment. We generally favor final, whenever possible.

**Review by korthout — COMMENTED** [review]
Starting to look good. Here are a few things I noted while looking into the escaping problem, but nothing major.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:158]: 🔧 That makes sense, please add this reasoning to the commit message

Comment (menski) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: I added it to the explanation in the commit message and this PR

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java
Comment (menski) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: I updated my commit to use a final variable

**Comment by menski** [comment]
@korthout I updated my code and adjusted the tests to expect the multiple escaping for the moment. Would be interested to know how you think we should continue with this? I also had to adjust another test `VariableMappingTransformerTest.java` which I think was just using the wrong input values, as it used static values instead of expressions, but maybe I also misunderstood it.

**Comment by korthout** [comment]
> I also had to adjust another test VariableMappingTransformerTest.java

Looks good to me 👍 

>how you think we should continue with this?

TBH, I think we can keep this problem out of scope. it's an already existing issue and not related to allowing static values in input mappings. It's just more visible with static values because they are always strings. 

Let's take the test you wrote, and change the input mappings from static, to string expressions:

```java
// Note that the string expressions require nested escapes just like you'd expect
private static final BpmnModelInstance WITH_EXPRESSION_INPUT_MAPPING =
      Bpmn.createExecutableProcess(PROCESS_ID)
          .startEvent()
          .serviceTask(
              "task",
              b ->
                  b.zeebeJobType("type")
                      .zeebeInputExpression("\"text\"", "text")
                      .zeebeInputExpression("\"https://my.api.com/{{orgId}}/{{userId}}\"", "url")
                      .zeebeInputExpression("\"my name is \\\"walt\\\"\"", "quotedText")
                      .zeebeInputExpression("\"{\\\"key\\\": \\\"value\\\"}\"", "object")
                      .zeebeInputExpression("\"[{\\\"key\\\": \\\"value\\\"}, 2, true]\"", "list")
                      .zeebeInputExpression("\"1\"", "number")
                      .zeebeInputExpression("\"true\"", "boolean"))
          .endEvent()
          .done();
```

and then with the test case:
```java
@Test
public void shouldApplyExpressionInputMapping() {
  // given
  ENGINE.deployment().withXmlResource(WITH_EXPRESSION_INPUT_MAPPING).deploy();
  final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();

  // when
  final Record<ProcessInstanceRecordValue> record =
      RecordingExporter.processInstanceRecords()
          .withElementId("task")
          .withIntent(ELEMENT_ACTIVATED)
          .withProcessInstanceKey(processInstanceKey)
          .getFirst();

  // then
  final Map<String, String> variables =
      ProcessInstances.getCurrentVariables(processInstanceKey, record.getPosition());

  // again note that the strings have nested escapes
  assertThat(variables)
      .contains(
          entry("text", "\"text\""),
          entry("url", "\"https://my.api.com/{{orgId}}/{{userId}}\""),
          entry("quotedText", "\"my name is \\\"walt\\\"\""),
          entry("object", "\"{\\\"key\\\": \\\"value\\\"}\""),
          entry("list", "\"[{\\\"key\\\": \\\"value\\\"}, 2, true]\""),
          entry("number", "\"1\""),
          entry("boolean", "\"true\""));
}
```

But this test fails as well (note the additional escapes again):
```
Expecting map:
  {"boolean"=""true"", "list"=""[{\\\"key\\\": \\\"value\\\"}, 2, true]"", "number"=""1"", "object"=""{\\\"key\\\": \\\"value\\\"}"", "quotedText"=""my name is \\\"walt\\\""", "text"=""text"", "url"=""https://my.api.com/{{orgId}}/{{userId}}""}
to contain:
  [MapEntry[key="text", value=""text""],
    MapEntry[key="url", value=""https://my.api.com/{{orgId}}/{{userId}}""],
    MapEntry[key="quotedText", value=""my name is \"walt\"""],
    MapEntry[key="object", value=""{\"key\": \"value\"}""],
    MapEntry[key="list", value=""[{\"key\": \"value\"}, 2, true]""],
    MapEntry[key="number", value=""1""],
    MapEntry[key="boolean", value=""true""]]
but could not find the following map entries:
  [MapEntry[key="quotedText", value=""my name is \"walt\"""],
    MapEntry[key="object", value=""{\"key\": \"value\"}""],
    MapEntry[key="list", value=""[{\"key\": \"value\"}, 2, true]""]]
```

Note, that the strings are correctly escaped when retrieved from the job (I used an activate job batch command here and called the `.getVariables()` method of the java-client which returns a `Map<String, Object>`).

Screen Shot 2022-03-02 at 15 07 59

Either, we have an actual bug with nested escaping in strings, or we just don't have the capability to test for it correctly. I think it is the latter. But either way, it seems to be an existing issue.

To come to a conclusion, we could argue whether we want to support static values in input mappings when it highlights this problem. I expect that users will define json objects through these static value input mappings. So, I suggest we test and see how Operate and Tasklist show these variables and then make a decision.

**Comment by menski** [comment]
> I expect that users will define json objects through these static value input mappings.

I'm not sure about this, because feel is actually the nicer thing to define json as you don't need escaping etc and can even use variable values, i.e. `= {key: variableValue, number: 123, list: ["a", "b", true]}`

> So, I suggest we test and see how Operate and Tasklist show these variables and then make a decision.

That is a really great idea, I modeled, deployed and started a process with the test string expressions. This is the outcome:

#### Operate

image

#### Tasklist

image

#### Worker 

```
➜ zbctl create worker test --handler 'python -m json.tool'
2022/03/09 22:59:51 Activated job 2251799813703250 with variables {"url":"https://my.api.com/{{orgId}}/{{userId}}","list":"[{\\\"key\\\":\\\"value\\\"}, 2, true]","text":"text","number":"123","object":"{\\\"key\\\": \\\"value\\\"}","boolean":"true","quotedText":"my name is \\\"walt\\\"."}
2022/03/09 22:59:51 Handler completed job 2251799813703250 with variables {
    "url": "https://my.api.com/{{orgId}}/{{userId}}",
    "list": "[{\\\"key\\\":\\\"value\\\"}, 2, true]",
    "text": "text",
    "number": "123",
    "object": "{\\\"key\\\": \\\"value\\\"}",
    "boolean": "true",
    "quotedText": "my name is \\\"walt\\\"."
}
```

#### Test process XML

[input-test.bpmn.txt](https://github.com/camunda-cloud/zeebe/files/8218455/input-test.bpmn.txt)


#### Summary

The double escaping is visible to the user, but maybe the assumption that users will use `"` and `\` in these fields is not correct there is no real issue. @korthout happy to get your input here

**Comment by korthout** [comment]
@menski Thanks for testing it

> feel is actually the nicer thing to define json as you don't need escaping etc and can even use variable values

👍 Ah yes, good point. I agree.

>The double escaping is visible to the user, but maybe the assumption that users will use `"` and `\` in these fields is not correct there is no real issue

Due to your point above, I think it is not an issue that blocks this PR.

- JSON objects can be used in FEEL context-expressions as a first-class citizen and so should be the preferred way to define JSON objects
- the double escaping behavior already exists for FEEL String-expressions. If we consider that behavior a bug and want to change it for static values, then we should also change it for FEEL expression results. If we want to do that, then we should consider it carefully. It is a behavioral change that affects user space and so may break existing process models. My preference goes out to leave it up to users to complain about this double escaping behavior before making such a change. But in any case, we should discuss this in a separate issue.

**Review by korthout — COMMENTED** [review]
I have 2 last comments

🔧 Let's add a test that showcases the double escaping behavior of FEEL string-expressions, similar to the static input mapping test.

❓ Do we also want to support static values for output mappings?

**Comment by menski** [comment]
> ￼ Let's add a test that showcases the double escaping behavior of FEEL string-expressions, similar to the static input mapping test.

I already extended my [test case](https://github.com/camunda-cloud/zeebe/pull/8844/files#diff-3e3efa3fb56b3f58856c74bc267e7bc8a4f24709d7755a680fc14858f0d96c8d) to have static values and expression to understand if there is any difference. Does that work for you or do you want something else?

> Do we also want to support static values for output mappings?

I don't see a use case right now for it and wouldn't see it in the scope of this issue.

**Review by korthout — APPROVED** [review]
Thanks @menski LGTM 👍

**Comment by menski** [comment]
Thanks for the help and review @korthout :1st_place_medal: 

bors merge

**Review by korthout — APPROVED** [review]
Let's try that again 🥈 LGTM

**Comment by menski** [comment]
bors merge

**Pull request merged**
